src/lib.rs

```diff
@@ -2,269 +2,234 @@
 //
 // SPDX-License-Identifier: MIT OR Apache-2.0
 
-/*! The library provide tools for handling compressed and archive files
-
-# Examples
-```no_run
-use compress_tools;
-
-let dir = tempfile::tempdir().unwrap();
-compress_tools::uncompress("tests/fixtures/tree.tar.gz", dir.path(), compress_tools::Kind::TarGZip).unwrap();
-```
-*/
-
-mod uncompress;
-
-pub use uncompress::{Error, Result};
-
-use std::path::Path;
-
-/// Type of compressed file or archive
-#[derive(Copy, Clone, Debug)]
-pub enum Kind {
-    TarGZip,
-    TarBZip2,
-    TarXz,
-    TarLZMA,
-    TarLZip,
-    Tar,
-    Zip,
-
-    GZip,
-    BZip2,
-    Xz,
-    LZip,
-    LZMA,
-}
-
-/// Uncompress a archive of known [Kind](Kind) pointed by source to target
-/// location for file
-pub fn uncompress<P1, P2>(source: P1, target: P2, kind: Kind) -> Result<()>
+pub mod error;
+pub mod ffi;
+pub(crate) mod util;
+
+pub use error as compress;
+use ffi::*;
+use std::{
+    ffi::{CStr, CString},
+    os::raw::c_void,
+};
+use util::*;
+
+/// Reads the input buffer and extract to another buffer.
+///
+/// # Arguments
+///
+/// * `source` - Reader.
+/// * `target` - Writer.
+///
+/// # Example
+///
+/// buffered_uncompress(std::io::Reader, std::io::Writer)?;
+///
+/// ```
+
+pub fn buffered_uncompress<R, W>(source: &mut R, target: &mut W) -> compress::Result<()>
 where
-    P1: AsRef<Path>,
-    P2: AsRef<Path>,
+    R: std::io::Read,
+    W: std::io::Write,
 {
-    let source = uncompress::Uncompress::new(source.as_ref());
-    let target = target.as_ref();
-    match kind {
-        // Single files
-        Kind::BZip2 => source.bz2().file(target),
-        Kind::GZip => source.gz().file(target),
-        Kind::LZMA => source.lzma().file(target),
-        Kind::LZip => source.lzip().file(target),
-        Kind::Xz => source.xz().file(target),
-
-        // Archive files
-        Kind::Tar => source.tar(target),
-        Kind::TarBZip2 => source.bz2().tar(target),
-        Kind::TarGZip => source.gz().tar(target),
-        Kind::TarLZMA => source.lzma().tar(target),
-        Kind::TarLZip => source.lzip().tar(target),
-        Kind::TarXz => source.xz().tar(target),
-        Kind::Zip => source.unzip(target),
-    }
+    let mut source_buffer = Vec::new();
+    let mut target_buffer = Vec::new();
+
+    source.read_exact(&mut source_buffer)?;
```

Comment:
This read of the source is unsound. `Vec::new()` generates an empty `Vec`, while `read_exact` would read exactly 0. It looks like nothing is being uncompressed here really.

The fact that the `uncompress_tar_2` test doesn't fail just shows it really needs to be reworked. I suggest you started by there to ensure that something is actually being uncompressed.

On a following note to fix the issue in hand, you gonna have to have a `Vec` set with some capacity and make multiple reads to the source until it has ended or the libarchive says that it should stop for some reason.